### PR TITLE
[DEV-2022] Add more fields when send SMS

### PR DESCRIPTION
**Pull request by chankiyu22**
refs DEV-2022

<details><summary>Previous preview</summary>

```diff
WARN[0031] skip sending SMS in development mode          app=my-app app_id=my-app body="Setup two-step verification on Authgear with this one-time password 353888.\n\nDon't share it with anyone.\n" language_tag=en logger=sms-client recipient=+85298765432 sender=Authgear template_name=messages/setup_secondary_oob_sms.txt template_variables="&{Authgear  353888  false accounts.portal.localhost:3100   +85298765432    }"
```
</details>

Updated.

```diff
-WARN[0031] skip sending SMS in development mode          app=my-app app_id=my-app body="Setup two-step verification on Authgear with this one-time password 353888.\n\nDon't share it with anyone.\n" language_tag=en logger=sms-client recipient=+85298765432 sender=Authgear template_name=messages/setup_secondary_oob_sms.txt template_variables="&{Authgear  353888  false accounts.portal.localhost:3100   +85298765432    }"
+WARN[0002] skip sending SMS in development mode          app=my-app app_id=my-app body="Setup two-step verification on Authgear with this one-time password 209812.\n\nDon't share it with anyone.\n" language_tag=en logger=sms-client message_type=setup-secondary-oob recipient=+85298765432 sender=Authgear template_name=setup_secondary_oob_sms.txt template_variables="&{Authgear  209812  false accounts.portal.localhost:3100   +85298765432    }"
```

### CODE REVIEW

**Comment by louischan-oursky** [comment]
Please rebase.

**Comment by chankiyu22** [comment]
@louischan-oursky rebased.

File: pkg/lib/infra/sms/variables.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package sms
+
+import (
+	"github.com/authgear/authgear-server/pkg/lib/translation"
+)
+
+type TemplateVariables struct {
+	AppName     string `json:"app_name"`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/variables.go:8]: Add omitempty to all fields.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
I found another change we need to address.

1. Currently there is `Name` in MessageSpec. The usage of `Name` is to construct a translation key. So please rename `Name` to `KeyComponentInTranslationJSON`.
2. Now that we know `KeyComponentInTranslationJSON` is not a general name. So you do not include it as extra fields.
3. There is a weird type nonblocking.MessageType. Temporarily delete it.
4. By deleting nonblocking.MessageType, you will encounter compilation errors. So our next step is to fix them. We want our code no longer depend on nonblocking.MessageType
5. nonblocking.MessageType ultimately will be written in the audit log. In addition to nonblocking.MessageType, we want to record the actual template name. The actual template name is `something.html` or `something.txt`.

File: pkg/lib/infra/sms/custom.go [hunk 2]
```diff
@@ -30,19 +30,38 @@ func NewCustomClient(c *config.CustomSMSProviderConfig, d SMSDenoHook, w SMSWebH
 }
 
 type SendSMSPayload struct {
-	To   string `json:"to"`
-	Body string `json:"body"`
+	To                string             `json:"to"`
+	Body              string             `json:"body"`
+	AppID             string             `json:"app_id"`
+	TemplateName      string             `json:"template_name"`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/custom.go:36]: MessageType string `json:"message_type"`
TemplateName string `json:"template_name"` (basename of the logic path)

File: pkg/lib/infra/sms/custom.go [hunk 3]
```diff
@@ -30,19 +30,38 @@ func NewCustomClient(c *config.CustomSMSProviderConfig, d SMSDenoHook, w SMSWebH
...
+	LanguageTag       string             `json:"preferred_language_tags"`
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/custom.go:37]: language_tag

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by chankiyu22** [comment]
@louischan-oursky Updated.

File: pkg/lib/translation/message_type.go [hunk 4]
```diff
@@ -1,21 +1,27 @@
-package otp
+package translation
```
Comment (louischan-oursky) [inline pkg/lib/translation/message_type.go:1]: This file should have name `message_declarations.go`

**Review by louischan-oursky — CHANGES_REQUESTED** [review]
1. nonblocking.MessageType -> translation.MessageType
2. MessageSpec.Name (string -> SpecName)
3. Add new field MessageSpec.MessageType -> MessageType
4. There is a function `(Form, otp.MessageType) -> (MessageSpec, nonblock.MessageType)`. After the change (Form, translation.MessageType) -> MessageSpec
5. Outsiders use MessageType only, they do not use SpecName.
6.

File: pkg/lib/authn/otp/sender.go [hunk 5]
```diff
@@ -33,9 +35,9 @@ type TranslationService interface {
 }
 
 type Sender interface {
-	PrepareEmail(email string, msgType nonblocking.MessageType) (*messaging.EmailMessage, error)
-	PrepareSMS(phoneNumber string, msgType nonblocking.MessageType) (*messaging.SMSMessage, error)
-	PrepareWhatsapp(phoneNumber string, msgType nonblocking.MessageType) (*messaging.WhatsappMessage, error)
+	PrepareEmail(email string, msgType string) (*messaging.EmailMessage, error)
+	PrepareSMS(phoneNumber string, msgType string) (*messaging.SMSMessage, error)
```
Comment (louischan-oursky) [inline pkg/lib/authn/otp/sender.go:39]: Change the type to translation.MessageType

File: pkg/lib/authn/otp/sender.go [hunk 6]
```diff
@@ -33,9 +35,9 @@ type TranslationService interface {
...
+	PrepareWhatsapp(phoneNumber string, msgType string) (*messaging.WhatsappMessage, error)
```
Comment (louischan-oursky) [inline pkg/lib/authn/otp/sender.go:40]: Change the type to translation.MessageType

File: pkg/lib/messaging/message_email.go [hunk 7]
```diff
@@ -12,7 +12,7 @@ type EmailMessage struct {
 	taskQueue task.Queue
 	events    EventService
 
-	Type nonblocking.MessageType
+	Type string
```
Comment (louischan-oursky) [inline pkg/lib/messaging/message_email.go:15]: translation.MessageType

File: pkg/lib/messaging/message_sms.go [hunk 8]
```diff
@@ -12,7 +12,7 @@ type SMSMessage struct {
 	taskQueue task.Queue
 	events    EventService
 
-	Type nonblocking.MessageType
+	Type string
```
Comment (louischan-oursky) [inline pkg/lib/messaging/message_sms.go:15]: translation.MessageType

File: pkg/lib/messaging/message_whatsapp.go [hunk 9]
```diff
@@ -11,7 +11,7 @@ type WhatsappMessage struct {
 	taskQueue task.Queue
 	events    EventService
 
-	Type         nonblocking.MessageType
+	Type         string
```
Comment (louischan-oursky) [inline pkg/lib/messaging/message_whatsapp.go:14]: translation.MessageType

File: pkg/lib/translation/message.go [hunk 10]
```diff
@@ -3,13 +3,9 @@ package translation
 import "github.com/authgear/authgear-server/pkg/util/template"
 
 type MessageSpec struct {
-	Name              string
+	Type              MessageType
```
Comment (louischan-oursky) [inline pkg/lib/translation/message.go:6]: Name SpecName

File: pkg/lib/translation/message_type.go [hunk 11]
```diff
@@ -1,21 +1,27 @@
...
 const (
-	MessageTypeVerification             MessageType = "verification"
-	MessageTypeSetupPrimaryOOB          MessageType = "setup-primary-oob"
-	MessageTypeSetupSecondaryOOB        MessageType = "setup-secondary-oob"
-	MessageTypeAuthenticatePrimaryOOB   MessageType = "authenticate-primary-oob"
-	MessageTypeAuthenticateSecondaryOOB MessageType = "authenticate-secondary-oob"
-	MessageTypeForgotPassword           MessageType = "forgot-password"
-	MessageTypeChangePassword           MessageType = "change-password"
-	MessageTypeWhatsappCode             MessageType = "whatsapp-code"
+	MessageTypeVerification                   MessageType = "verification"
```
Comment (louischan-oursky) [inline pkg/lib/translation/message_type.go:10]: SpecNames

File: pkg/lib/translation/message_type.go [hunk 12]
```diff
@@ -60,73 +66,89 @@ var (
...
+
+	TemplateMessageSendPasswordToNewUserTXT       = template.RegisterMessagePlainText("messages/send_password_to_new_user_email.txt")
+	TemplateMessageSendPasswordToNewUserEmailHTML = template.RegisterMessageHTML("messages/send_password_to_new_user_email.html")
 )
 
 var (
-	messageVerification = &translation.MessageSpec{
-		Name:              "verification",
+	MessageVerification = &MessageSpec{
+		Type:              MessageTypeVerification,
```
Comment (louischan-oursky) [inline pkg/lib/translation/message_type.go:79]: Ensure they have SpecName and Type.

File: pkg/lib/translation/service.go [hunk 13]
```diff
@@ -66,29 +66,29 @@ func (s *Service) GetSenderForTestEmail() (sender string, err error) {
 	return
 }
 
-func (s *Service) emailMessageHeader(name string, variables *PreparedTemplateVariables) (sender, replyTo, subject string, err error) {
+func (s *Service) emailMessageHeader(name MessageType, variables *PreparedTemplateVariables) (sender, replyTo, subject string, err error) {
```
Comment (louischan-oursky) [inline pkg/lib/translation/service.go:69]: name SpecName

File: pkg/lib/translation/service.go [hunk 14]
```diff
@@ -140,13 +140,13 @@ func (s *Service) EmailMessageData(msg *MessageSpec, variables *PartialTemplateV
 	}, nil
 }
 
-func (s *Service) smsMessageHeader(name string, variables *PreparedTemplateVariables) (sender string, err error) {
+func (s *Service) smsMessageHeader(name MessageType, variables *PreparedTemplateVariables) (sender string, err error) {
```
Comment (louischan-oursky) [inline pkg/lib/translation/service.go:143]: name SpecName

File: pkg/lib/translation/service_test.go [hunk 15]
```diff
@@ -28,7 +28,7 @@ func TestService(t *testing.T) {
 		var TemplateMessageWhatsappTXT = template.RegisterMessagePlainText("messages/whatsapp.txt")
 
 		var messageSpec = &translation.MessageSpec{
-			Name:              "test",
+			Type:              "test",
```
Comment (louischan-oursky) [inline pkg/lib/translation/service_test.go:31]: Update accordingly.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by chankiyu22** [comment]
@louischan-oursky updated.

File: pkg/lib/infra/sms/client.go [hunk 16]
```diff
@@ -22,7 +22,16 @@ type SendOptions struct {
...
+	Send(
+		from string,
+		to string,
+		body string,
+		appID string,
+		messageType string,
+		templateName string,
+		languageTag string,
+		variables *TemplateVariables,
+	) error
```
Comment (louischan-oursky) [inline pkg/lib/infra/sms/client.go:34]: Take SendOptions directly.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by chankiyu22** [comment]
@louischan-oursky updated.

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**